python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_smart_wallet_provider.py

```diff
@@ -0,0 +1,396 @@
+from typing import Any
+
+from cdp import CdpClient
+from cdp.evm_call_types import EncodedCall
+from eth_account import Account
+from pydantic import BaseModel, Field
+from web3 import Web3
+from web3.types import BlockIdentifier, ChecksumAddress, HexStr, TxParams
+
+from ..network import NETWORK_ID_TO_CHAIN, Network
+from .evm_wallet_provider import EvmGasConfig, EvmWalletProvider
+
+
+class CdpEvmSmartWalletProviderConfig(BaseModel):
+    """Configuration options for CDP EVM Smart Wallet provider."""
+
+    api_key_id: str | None = Field(None, description="The CDP API key ID")
+    api_key_secret: str | None = Field(None, description="The CDP API secret")
+    wallet_secret: str | None = Field(None, description="The CDP wallet secret")
+    network_id: str | None = Field(None, description="The network id")
+    address: str | None = Field(None, description="The smart wallet address to use")
+    owner: str | None = Field(
+        None, description="The owner's private key or CDP server wallet address"
+    )
+    idempotency_key: str | None = Field(None, description="The idempotency key for wallet creation")
+    gas: EvmGasConfig | None = Field(None, description="Gas configuration settings")
+    paymaster_url: str | None = Field(
+        None, description="Optional paymaster URL for gasless transactions"
+    )
+
+
+class CdpEvmSmartWalletProvider(EvmWalletProvider):
+    """A wallet provider that uses the CDP EVM Smart Account SDK."""
+
+    def __init__(self, config: CdpEvmSmartWalletProviderConfig):
+        """Initialize CDP EVM Smart Wallet provider.
+
+        Args:
+            config (CdpEvmSmartWalletProviderConfig | None): Configuration options for the CDP provider. If not provided,
+                   will attempt to configure from environment variables.
+
+        Raises:
+            ValueError: If required configuration is missing or initialization fails
+
+        """
+        try:
+            self._api_key_id = config.api_key_id or os.getenv("CDP_API_KEY_ID")
+            self._api_key_secret = config.api_key_secret or os.getenv("CDP_API_KEY_SECRET")
+            self._wallet_secret = config.wallet_secret or os.getenv("CDP_WALLET_SECRET")
+            self._paymaster_url = config.paymaster_url
+            self._address = config.address
+            owner_address_or_private_key = config.owner or os.getenv("OWNER")
+
+            if not self._api_key_id or not self._api_key_secret or not self._wallet_secret:
+                raise ValueError(
+                    "Missing required environment variables. CDP_API_KEY_ID, CDP_API_KEY_SECRET, CDP_WALLET_SECRET are required."
+                )
+
+            if not owner_address_or_private_key:
+                raise ValueError("Owner private key or CDP server wallet address is required")
+
+            network_id = config.network_id or os.getenv("NETWORK_ID", "base-sepolia")
+            self._idempotency_key = config.idempotency_key or os.getenv("IDEMPOTENCY_KEY")
```

Comment:
I don't think this is actually used anywhere in this class currently

Reply:
You are right, it is not. There is currently no way to set a idempotency key when creating a smart wallet with the `cdp-sdk`, however I assume there will be in the future? If not, I'll delete this entirely